extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java

```diff
@@ -122,13 +137,13 @@ public QueryDefinition makeQueryDefinition(
       queryDefBuilder.add(
           StageDefinition.builder(firstStageNumber)
                          .inputs(new StageInputSpec(firstStageNumber - 1))
-                         .signature(rowSignature)
+                         .signature(finalWindowStageRowSignature)
                          .maxWorkerCount(maxWorkerCount)
-                         .shuffleSpec(null)
+                         .shuffleSpec(finalWindowStageShuffleSpec)
```

Comment:
This is being done only for the last window stage. If we have an empty over() clause present in the window function, we create a single window stage instead of separate stages.

With that said, I've pushed a change to add the virtual columns also only for the last window stage. Hope that works!

Reply:
From what I can understand, you are adding logic which uses the segment granularity to decide the shuffling. However, it would be possible to have more than one window stage in a query, and the query does not need to be followed by a segment generation stage. You can add a limit to a window query to see this behavior. I am wondering if this case is covered as well.

Reply:
@adarshsanjeev We're using the resultShuffleSpecFactory only for the last window stage. I'm not sure I fully understand the comment.

For the final stage, we have the following cases for segmentGranularity:
1. Granularities.ALL: We set the shuffle spec to MixShuffleSpec for the final window stage in this case.
2. All other types: We create the shuffle spec using resultShuffleSpecFactory for the final window stage in this case.

I tried running the following query as well, as suggested (IIUC):
```sql
select countryName, cityName, channel, 
row_number() over (partition by countryName order by countryName, cityName, channel) as c1, 
row_number() over (partition by cityName order by countryName, cityName, channel) as c2
from wikipedia
where countryName in ('Republic of Korea', 'Austria')
group by countryName, cityName, channel
limit 7
```

This query gives correct results.

Reply:
For the above query, these are the stages we get:
```
stageDefinition.getProcessorFactory() = org.apache.druid.msq.querykit.groupby.GroupByPreShuffleFrameProcessorFactory@737a033
stageDefinition.getSignature() = {d0:STRING, d1:STRING, d2:STRING}
stageDefinition.getStageNumber() = 0
stageDefinition.getShuffleSpec() = MaxCountShuffleSpec{clusterBy=ClusterBy{columns=[d0 ASCENDING, d1 ASCENDING, d2 ASCENDING], bucketByCount=0}, partitions=1, aggregate=true}


stageDefinition.getProcessorFactory() = org.apache.druid.msq.querykit.groupby.GroupByPostShuffleFrameProcessorFactory@3b66dec
stageDefinition.getSignature() = {d0:STRING, d1:STRING, d2:STRING}
stageDefinition.getStageNumber() = 1
stageDefinition.getShuffleSpec() = HashShuffleSpec{clusterBy=ClusterBy{columns=[d0 ASCENDING], bucketByCount=0}, numPartitions=1}


stageDefinition.getProcessorFactory() = org.apache.druid.msq.querykit.WindowOperatorQueryFrameProcessorFactory@d45add62
stageDefinition.getSignature() = {d1:STRING, d0:STRING, d2:STRING, w0:LONG}
stageDefinition.getStageNumber() = 2
stageDefinition.getShuffleSpec() = HashShuffleSpec{clusterBy=ClusterBy{columns=[d1 ASCENDING], bucketByCount=0}, numPartitions=1}


stageDefinition.getProcessorFactory() = org.apache.druid.msq.querykit.WindowOperatorQueryFrameProcessorFactory@643913d9
stageDefinition.getSignature() = {d0:STRING, d1:STRING, d2:STRING, w0:LONG, w1:LONG}
stageDefinition.getStageNumber() = 3
stageDefinition.getShuffleSpec() = MixShuffleSpec{}


stageDefinition.getProcessorFactory() = org.apache.druid.msq.querykit.scan.ScanQueryFrameProcessorFactory@687279fb
stageDefinition.getSignature() = {__boost:LONG, d0:STRING, d1:STRING, d2:STRING, w0:LONG, w1:LONG}
stageDefinition.getStageNumber() = 4
stageDefinition.getShuffleSpec() = MixShuffleSpec{}


stageDefinition.getProcessorFactory() = org.apache.druid.msq.querykit.common.OffsetLimitFrameProcessorFactory@3c8
stageDefinition.getSignature() = {__boost:LONG, d0:STRING, d1:STRING, d2:STRING, w0:LONG, w1:LONG}
stageDefinition.getStageNumber() = 5
stageDefinition.getShuffleSpec() = MaxCountShuffleSpec{clusterBy=ClusterBy{columns=[__boost ASCENDING], bucketByCount=0}, partitions=1, aggregate=false}
```

Reply:
I meant with INSERT, but I think that case should also work. I wasn't sure initially.